Allowing avahi-observer/control slots from app snap also on classic

kubiko opened the pull request:
On classis assumption is avahi is installed as deb and provided by default
however this if often not case for server images

From avahi snap metrics, this assumption turned up to be flawed.
Since avahi socked is now moved to same place as with deb, avahi snap does not have sufficient permissions to run in classic systems

Signed-off-by: Ondrej Kubik <ondrej.kubik@canonical.com>

Thanks for helping us make a better snapd!
Have you signed the [license agreement](https://www.ubuntu.com/legal/contributors) and read the [contribution guide](https://github.com/snapcore/snapd/blob/master/CONTRIBUTING.md)?

kubiko commented:
For reference here is forum post about fail
https://forum.snapcraft.io/t/avahi-daemon-wont-start/8990/2
For reference, from snap metrics I can see over 90% of installs being on classic system (which is now broken since we moved socket to more predictable place)

jdstrand commented on interfaces/builtin/avahi_control.go:
```diff
@@ -120,9 +120,12 @@ func (iface *avahiControlInterface) AppArmorConnectedPlug(spec *apparmor.Specifi
 		// of the OS snap and will run unconfined.
-		new = "unconfined"
+		new = slotAppLabelExpr(slot)
+		if new == "\"snap.core.*\"" {
```
This doesn't look like it would work for core16/core18-based systems...

jdstrand commented on interfaces/builtin/avahi_control.go:
```diff
@@ -120,9 +120,12 @@ func (iface *avahiControlInterface) AppArmorConnectedPlug(spec *apparmor.Specifi
+		}
 	} else {
 		new = slotAppLabelExpr(slot)
 	}
```
This if/else could be reorganized to be a little more clean I think with:
```
new = slotAppLabelExpr(slot)
if release.OnClassic && new == "\"snap.core.*\"" { # fix for core16/core18
	new = "unconfined"
}
```

jdstrand commented on interfaces/builtin/avahi_control.go:
```diff
@@ -135,33 +138,27 @@ func (iface *avahiControlInterface) AppArmorConnectedPlug(spec *apparmor.Specifi
-	}
+	// NOTE: this is using avahi-observe permanent slot as it contains
+	// base declarations for running as the avahi service.
+	spec.AddSnippet(avahiObservePermanentSlotAppArmor)
```
The PermanentSlot policy should not be applied when the slotting snap is core, core18, etc.

jdstrand commented on interfaces/builtin/avahi_control.go:
```diff
@@ -135,33 +138,27 @@ func (iface *avahiControlInterface) AppArmorConnectedPlug(spec *apparmor.Specifi
+	snippet := strings.Replace(avahiObserveConnectedSlotAppArmor, old, new, -1)
+	spec.AddSnippet(snippet)
+	snippet = strings.Replace(avahiControlConnectedSlotAppArmor, old, new, -1)
+	spec.AddSnippet(snippet)
```
Likewise, the ConnectedSlot policy should not be applied when the slotting snap is core, core18, etc.

jdstrand commented on interfaces/builtin/avahi_control.go:
```diff
@@ -135,33 +138,27 @@ func (iface *avahiControlInterface) AppArmorConnectedPlug(spec *apparmor.Specifi
-	}
+	// NOTE: this is using avahi-observe permanent slot as it contains
+	// base declarations for running as the avahi service.
+	spec.AddSnippet(avahiObservePermanentSlotDBus)
```
Likewise, the PermanentSlot policy for dbus should not be applied when the slotting snap is core, core18, etc. The deb/rpm will already have dbus policy on the system and we shouldn't try to override that.

jdstrand commented on interfaces/builtin/avahi_observe.go:
```diff
@@ -431,9 +431,12 @@ func (iface *avahiObserveInterface) AppArmorConnectedPlug(spec *apparmor.Specifi
+		new = slotAppLabelExpr(slot)
+		if new == "\"snap.core.*\"" {
+			new = "unconfined"
+		}
```
See comments for avahi_control.go.

jdstrand commented on interfaces/builtin/avahi_observe.go:
```diff
@@ -443,26 +446,20 @@ func (iface *avahiObserveInterface) AppArmorConnectedPlug(spec *apparmor.Specifi
-	if !release.OnClassic {
-		spec.AddSnippet(avahiObservePermanentSlotAppArmor)
-	}
+	spec.AddSnippet(avahiObservePermanentSlotAppArmor)
```
See comments for avahi_control.go.

jdstrand commented on interfaces/builtin/avahi_observe.go:
```diff
@@ -443,26 +446,20 @@ func (iface *avahiObserveInterface) AppArmorConnectedPlug(spec *apparmor.Specifi
+	old := "###PLUG_SECURITY_TAGS###"
+	new := plugAppLabelExpr(plug)
+	snippet := strings.Replace(avahiObserveConnectedSlotAppArmor, old, new, -1)
+	spec.AddSnippet(snippet)
```
See comments for avahi_control.go.

jdstrand commented on interfaces/builtin/avahi_observe.go:
```diff
@@ -443,26 +446,20 @@ func (iface *avahiObserveInterface) AppArmorConnectedPlug(spec *apparmor.Specifi
-	if !release.OnClassic {
-		spec.AddSnippet(avahiObservePermanentSlotDBus)
-	}
+	spec.AddSnippet(avahiObservePermanentSlotDBus)
```
See comments for avahi_control.go.

om26er commented:
@kubiko Do you think you can revert the avahi snap to create the sock in a "non-standard" directory for now ? The snap is currently broken on classic, which as a result breaks my project that is supposed to be discoverable on the local network (both UbuntuCore and Classic)

kubiko commented:
@jdstrand thanks for great comments and chat about right way to handle this. I realised previous implementation wasn't really approaching it right way, it went from one extreme to another.
I added now helper functions to determine this specific case when slot is implicit on classic, but could be also provided by app snap. Just as you suggested.
I also moved from using snap names and rule to determine slot provider and instead used snap type, this seems more reliable approach going forward

kubiko commented:
@om26er let's give it second try before reverting

kubiko commented:
@stolowski can you please check chase I did in `interfaces/builtin/utils_test.go`
I updated slotSnapd type from `snap.TypeApp` -> `snap.TypeApp`
```- slotSnapd:  &snap.SlotInfo{Snap: &snap.Info{Type: snap.TypeApp, SuggestedName: "snapd"}},
+ slotSnapd:    &snap.SlotInfo{Snap: &snap.Info{Type: snap.TypeSnapd, SuggestedName: "snapd"}},

stolowski commented on interfaces/builtin/utils_test.go:
```diff
@@ -32,19 +32,25 @@ import (
+	iface:        &ifacetest.TestInterface{InterfaceName: "iface"},
+	slotOS:       &snap.SlotInfo{Snap: &snap.Info{Type: snap.TypeOS}},
+	slotApp:      &snap.SlotInfo{Snap: &snap.Info{Type: snap.TypeApp}},
+	slotSnapd:    &snap.SlotInfo{Snap: &snap.Info{Type: snap.TypeSnapd, SuggestedName: "snapd"}},
```
This change is fine although although technically it makes no difference for sanitizeSlotReservedForOS() as at the moment it is based on snap name check. But it's the right thing to do. Thanks.

mvo5 commented on interfaces/builtin/avahi_control.go:
```diff
@@ -119,9 +118,10 @@ func (iface *avahiControlInterface) StaticInfo() interfaces.StaticInfo {
-		// If we're running on classic Avahi will be part
-		// of the OS snap and will run unconfined.
+	// If we're running on classic Avahi can be installed as snap or be part
+	// of the OS and it will run unconfined.
```
Does this comment need to be expanded? Something like:
```
	// If we're running on classic Avahi can be installed as snap or be part
	// of the OS and it will run unconfined *or* it can be installed as a snap
        // with the normal confined snap label
```

kubiko replied:
updated

mvo5 commented on interfaces/builtin/avahi_control.go:
```diff
@@ -119,9 +118,10 @@ func (iface *avahiControlInterface) StaticInfo() interfaces.StaticInfo {
-		// of the OS snap and will run unconfined.
+	// If we're running on classic Avahi can be installed as snap or be part
+	// of the OS and it will run unconfined.
+	if isConnectedSlotSystemSlot(slot) {
```
(nitpick) I don't really like the name - at the same time I'm not great with names. Maybe some suggestions:
- implicitSystemSlot(slot)
- ... ?

kubiko replied:
yeah same here I don't know what's best name here I changed it already good 5 times :)
- implicitSystemSlot
- connectedImplicitSystemSlot ?

kubiko replied:
I like those names, updated based on proposal from Jamie, thank you @jdstrand

mvo5 commented on interfaces/builtin/avahi_control.go:
```diff
@@ -135,7 +135,9 @@ func (iface *avahiControlInterface) AppArmorConnectedPlug(spec *apparmor.Specifi
-	if !release.OnClassic {
+	// only apply slot snipped when running as application snap
+	// on classic, slot side can be system or application
+	if !isSlotSystemSlot(slot) {
```
Same as above, maybe:
- implicitSystemSlot(slot) 
- ?
(also not great but…)

kubiko replied:
Used proposal from Jamie, thank you :)

mvo5 commented on interfaces/builtin/utils.go:
```diff
@@ -96,3 +97,26 @@ func sanitizeSlotReservedForOSOrApp(iface interfaces.Interface, slot *snap.SlotI
+// - slot owned by the system (core,snapd snap)  usually requires no action
+// - slot owned by application snap typically requires rules update
+func isSlotSystemSlot(slot *snap.SlotInfo) bool {
+	if release.OnClassic &&
```
Do we actually need this? The function today is used in a way that we can't have "avahi" on !classic, right?

kubiko replied:
we need it to determine how is avahi installed.
Avahi can be on classic installed as deb or as snap. And clients should be able to connect to either one.
Trick is when installing avahi on classic as snap, we need to give it right permissions, but without giving those permissions to core snap in case avahi is installed as deb....
So this function should determine if slot in question is system slot ( avahi installed as deb) or not (avahi installed as snap)
We do not have this case yet, but I can image blues, network manager being another good examples....

mvo5 replied:
Yeah, what I mean I that I wonder if we need the explicit check for `release.OnClassic` here? Given the current uses/use-cases for this its fine mostly wondering if at some point down the line we write code like "implicitSystemPermanentSlot(kernelModuleCOntrolSlot)" and wonder why we get different results on core vs classic. Maybe @pedronis has an opinion?

mvo5 commented on interfaces/builtin/utils.go:
```diff
@@ -96,3 +97,26 @@ func sanitizeSlotReservedForOSOrApp(iface interfaces.Interface, slot *snap.SlotI
+// - slot owned by application snap typically requires rules update
+func isSlotSystemSlot(slot *snap.SlotInfo) bool {
+	if release.OnClassic &&
+		(slot.Snap.Type == snap.TypeOS || slot.Snap.Type == snap.TypeSnapd) {
```
This is good but probably needs to include a check for the snap name "snapd" - we don't use the snapd type yet (mostly because in the past the store was not ready and now I think it is but we dropped the ball).

kubiko replied:
@mvo5 instead of adding name check to many different places, wouldn't it be cleaner to simply choose correct enum when populating snap info? So we have name check only in one place?

kubiko commented on interfaces/builtin/avahi_control.go:
```diff
@@ -119,9 +118,11 @@ func (iface *avahiControlInterface) StaticInfo() interfaces.StaticInfo {
+	// with the normal confined snap label
+	if isConnectedSlotSystemSlot(slot) {
+		// avahi is part of the OS and runs unconfined.
 		new = "unconfined"
```
I notice those apparmor profiles in avahi deb package, so let's add it here
code updated

kubiko commented on interfaces/builtin/avahi_control.go:
```diff
@@ -119,9 +118,11 @@ func (iface *avahiControlInterface) StaticInfo() interfaces.StaticInfo {
-		// of the OS snap and will run unconfined.
+	// If we're running on classic Avahi can be installed as snap or be part
+	// of the OS and it will run unconfined *or* it can be installed as a snap
+	// with the normal confined snap label
```
Updated with proposed comment

kubiko commented on interfaces/builtin/avahi_control.go:
```diff
@@ -135,7 +136,9 @@ func (iface *avahiControlInterface) AppArmorConnectedPlug(spec *apparmor.Specifi
 
 func (iface *avahiControlInterface) AppArmorPermanentSlot(spec *apparmor.Specification, slot *snap.SlotInfo) error {
-	if !release.OnClassic {
+	// only apply slot snipped when running as application snap
```
ups, good find!
Fixed

kubiko commented on interfaces/builtin/avahi_observe.go:
```diff
@@ -430,9 +429,10 @@ func (iface *avahiObserveInterface) StaticInfo() interfaces.StaticInfo {
+	// of the OS and it will run unconfined.
+	if isConnectedSlotSystemSlot(slot) {
+		// avahi is part of the OS and runs unconfined.
 		new = "unconfined"
```
updated here as well

mvo5 commented:
The test failures look real:
```
AppArmor parser error for /var/lib/snapd/apparmor/profiles/snap.test-snapd-policy-app-consumer.avahi-control in /var/lib/snapd/apparmor/profiles/snap.test-snapd-policy-app-consumer.avahi-control at line 492: syntax error, unexpected TOK_CONDID, expecting TOK_EQUALS or TOK_IN
```

jdstrand commented on interfaces/builtin/avahi_observe.go:
```diff
@@ -430,10 +429,12 @@ func (iface *avahiObserveInterface) StaticInfo() interfaces.StaticInfo {
+	// label like it would when installed as a snap.
+	if implicitSystemConnectedSlot(slot) {
+		// avahi from the OS is typically unconfined but known to sometimes be confined
+		new = "{unconfined,/usr/sbin/avahi-daemon}"
```
Please change this to:
```
		new = "\"{unconfined,/usr/sbin/avahi-daemon}\""
```

kubiko replied:
Thank you @jdstrand 
This is now changed and tested. Now it works when I use avahi-client and connect its plug to implicit slot

jdstrand commented on interfaces/builtin/avahi_control.go:
```diff
@@ -119,10 +118,12 @@ func (iface *avahiControlInterface) StaticInfo() interfaces.StaticInfo {
+	// label like it would when installed as a snap.
+	if implicitSystemConnectedSlot(slot) {
+		// avahi from the OS is typically unconfined but known to sometimes be confined
+		new = "{unconfined,/usr/sbin/avahi-daemon}"
```
Please change this to:
```
		new = "\"{unconfined,/usr/sbin/avahi-daemon}\""
```

kubiko replied:
Thank you @jdstrand 
This is now changed and tested. Now it works when I use avahi-client and connect its plug to implicit slot

jdstrand commented:
Assuming manual testing shows this to still work, LGTM.

kubiko commented:
> Assuming manual testing shows this to still work, LGTM.

@jdstrand yes, I have tested it also manually on classic and core system. Running as expected
